src/server/coprocessor/executor/index_scan.rs

```diff
@@ -0,0 +1,214 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+use server::coprocessor::endpoint::prefix_next;
+use server::coprocessor::Result;
+use tipb::executor::IndexScan;
+use tipb::schema::ColumnInfo;
+use kvproto::coprocessor::KeyRange;
+use storage::{SnapshotStore, Statistics};
+use util::codec::{table, datum, mysql};
+use byteorder::{BigEndian, ReadBytesExt};
+use super::{Executor, Row};
+use super::base_scanner::BaseScanner;
+
+struct IndexScanExec<'a> {
+    meta: IndexScan,
+    col_ids: Vec<i64>,
+    cursor: usize,
+    key_ranges: Vec<KeyRange>,
+    scanner: BaseScanner<'a>,
+    pk_col: Option<ColumnInfo>,
+}
+
+impl<'a> IndexScanExec<'a> {
+    #[allow(dead_code)] //TODO:remove it
+    pub fn new(meta: IndexScan,
+               key_ranges: Vec<KeyRange>,
+               store: SnapshotStore<'a>,
+               statistics: &'a mut Statistics)
+               -> IndexScanExec<'a> {
+        let mut pk_col = None;
+        let col_ids = meta.get_columns()
+            .iter()
+            .filter(|c| {
+                if c.get_pk_handle() {
+                    pk_col = Some((*c).clone());
+                    false
+                } else {
+                    true
+                }
+            })
+            .map(|c| c.get_column_id())
+            .collect();
+        let scanner = BaseScanner::new(meta.get_desc(), false, store, statistics);
+        IndexScanExec {
+            meta: meta,
+            col_ids: col_ids,
+            scanner: scanner,
+            key_ranges: key_ranges,
+            cursor: 0,
+            pk_col: pk_col,
+        }
+    }
+
+    pub fn get_row_from_range(&mut self) -> Result<Option<Row>> {
+        let range = &self.key_ranges[self.cursor];
+        if range.get_start() > range.get_end() {
+            return Ok(None);
+        }
+        let kv = try!(self.scanner.get_row_from_range(range));
+        let (key, value) = match kv {
+            Some((key, value)) => (key, value),
```

Comment:
Try to avoid shadowing outer variables for easier understanding, even in simple case. It's better to write it like:
```
let (key, value) = match kv {
Some((k, k)) => (k, v),
...
```